scripts: footprint: Add a plot script for size reports

pdgendt opened the pull request:
Add a helper script that can take in a size report in json and render it in a sunburst chart.

UPDATE: Added CMake targets too

For example:

```shell
$ west build -p -b mimxrt1060_evk --shield rk043fn66hs_ctg samples/modules/lvgl/demos/ -t ram_plot
```

Opens a webpage, you can click through the different slices for more details, and hover to get more details:
<img width="1566" height="1040" alt="image" src="https://github.com/user-attachments/assets/27f289a7-1c5b-44f8-afff-cc34ba3ff032" />

pdgendt committed:
scripts: footprint: Add a plot script for size reports

Add a helper script that can take in a size report in json and render it in
a sunburst chart.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt commented:
> Would be even nicer to have this as a new option right in the ram_report / rom_report command, no?

Maybe, but AFAICS it wasn't possible to pass parameters to `west build -t ram_report`?

dsseng commented on scripts/footprint/plot.py:
```diff
@@ -0,0 +1,113 @@
+   ./scripts/footprint/plot.py build/ram.json
+
+Requires plotly to be installed, for example with pip:
+   pip install plotly
```
Maybe add this to `scripts/requirements-extras.txt` so that most users pick this dependency up by default?

pdgendt replied:
I preferred to keep the script and its dependency as standalone. The error message should be clear enough IMO, but if others share your opinion, we could add it.

dsseng replied:
Yes, it is pretty clear, and I'm not sure what the process of introducing a dependency is. Let's wait for others

pdgendt replied:
Update: added CMake targets too, and because of this added the dependency as you suggested.

dsseng reviewed (commented):
Great addition!

pdgendt committed:
cmake: reports: Add ram/rom plot targets

Add a plot variant for every ram/rom report target.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

dsseng commented on scripts/requirements-extras.txt:
```diff
@@ -26,3 +26,6 @@ PyGithub
 
 # used to generate devicetree dependency graphs
 graphviz
+
+# used by scripts/footprint/plot.py for generating plots of size reports
+plotly
```
Nit: move up to line 7, as these are similar tools

kartben commented on scripts/footprint/plot.py:
```diff
@@ -0,0 +1,99 @@
+        fig.write_html(args.html)
+        return
+
+    fig.show()
```
> Now running `west build -t rom_plot` would open your browser, which is nice, I think.

It is :) Could you please add some basic docs too, mainly so that folks know this exists?
https://docs.zephyrproject.org/latest/develop/optimizations/tools.html

pdgendt replied:
Added a `doc` commit, PTAL.

pdgendt committed:
doc: develop: optimizations: tools: Add ram_plot/rom_plot targets

Add information about the ram_plot and rom_plot build targets to generate
memory usage reports using sunburst charts.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt commented:
> * on a headless system I am getting a black screen with a tiny message saying: "Request sent". I suggest to make this behave the same way as puncover, i.e. print more details on the screen before auto opening, so if something fails, the information is printed out.,
> * does this support opening on a custom server/port? Maybe this will work similar to puncover and we could provide those via env variables.

Will check

> * once you clicked on something (drill down), there is no way to go back?

You need to click the center to go up.

pdgendt commented:
@nashif can you revisit this?

nashif commented:
btw, it would be nice to have this type of visualization integrated into puncover, then we will have one place to look at footprint and have more details where needed.....

nashif reviewed (dismissed):
still not great, and getting a blank screen immediately after the added message, but whatever, lets see how this goes :)

kartben reviewed (commented):
@pdgendt is a dependency missing?

```
  File "/Users/kartben/zephyrproject/zephyr/scripts/footprint/plot.py", line 114, in <module>
    main()
    ~~~~^^
  File "/Users/kartben/zephyrproject/zephyr/scripts/footprint/plot.py", line 110, in main
    fig.show()
    ~~~~~~~~^^
  File "/Users/kartben/zephyrproject/.venv/lib/python3.13/site-packages/plotly/basedatatypes.py", line 3420, in show
    return pio.show(self, *args, **kwargs)
           ~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/kartben/zephyrproject/.venv/lib/python3.13/site-packages/plotly/io/_renderers.py", line 415, in show
    raise ValueError(
        "Mime type rendering requires nbformat>=4.2.0 but it is not installed"
    )
ValueError: Mime type rendering requires nbformat>=4.2.0 but it is not installed
```


This is on OSX. Installing said dependency get past the error but then the `$ west build -p -b mimxrt1060_evk --shield rk043fn66hs_ctg samples/modules/lvgl/demos/ -t ram_plot` spits the HTML content on stdout?

pdgendt commented:
> This is on OSX. Installing said dependency get past the error but then the `$ west build -p -b mimxrt1060_evk --shield rk043fn66hs_ctg samples/modules/lvgl/demos/ -t ram_plot` spits the HTML content on stdout?

I didn't have this issue on my mac, but setting the browser as the renderer does sound sane.

kartben commented on doc/develop/optimizations/ram_plot.png:
Are these based on current status of the PR? I have zero color variation myself, which makes me a sad puppy :(

<img width="1070" height="941" alt="Image" src="https://github.com/user-attachments/assets/7ed1807e-e51d-41a3-ba21-a1e8b67fa849" />

kartben replied:
ah yes, cool, works great with reel_boards builds indeed - thanks!
